File: TP1/étudiant.py
from .personne import Personne

class Etudiant(Personne):
    def __init__(self, nom: str, age: int, numero_etudiant: str, moyenne: float):
        super().__init__(nom, age)
        self._numero_etudiant = numero_etudiant
        self.moyenne = moyenne
        self.liste_cours = []

    @property
    def numero_etudiant(self):
        return self._numero_etudiant

    @property
    def moyenne(self):
        return self._moyenne

    @moyenne.setter
    def moyenne(self, value):
        if not (0.0 <= value <= 20.0):
            raise ValueError("Moyenne invalide")
        self._moyenne = value

    def ajouter_cours(self, cours):
        self.liste_cours.append(cours)

    def afficher_details(self) -> str:
        base = super().afficher_details()
        return f"{base} (Etudiant ID: {self.numero_etudiant}, Moyenne: {self.moyenne})"

File: TP1/étudiant.java
import java.util.ArrayList;
import java.util.List;

public class Etudiant extends Personne {
    private String numEtudiant;
    private double moyenne;
    private List<Cours> listeCours = new ArrayList<>();

    public Etudiant(String nom, int age, String numEtudiant, double moyenne) {
        super(nom, age);
        this.numEtudiant = numEtudiant;
        setMoyenne(moyenne);
    }

    public String getNumEtudiant() { return numEtudiant; }
    
    public double getMoyenne() { return moyenne; }
    public void setMoyenne(double moyenne) {
        if (moyenne < 0 || moyenne > 20) throw new IllegalArgumentException();
        this.moyenne = moyenne;
    }

    public void ajouterCours(Cours c) {
        this.listeCours.add(c);
    }

    @Override
    public String afficherDetails() {
        return super.afficherDetails() + " (Etudiant ID: " + numEtudiant + ", Moyenne: " + moyenne + ")";
    }
}
